presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java

```diff
@@ -92,8 +91,8 @@ public BuiltInScalarFunctionImplementation specialize(BoundVariables boundVariab
                         TypeSignatureParameter.of(toKeyType.getTypeSignature()),
                         TypeSignatureParameter.of(toValueType.getTypeSignature())));
 
-        MethodHandle keyProcessor = buildProcessor(functionAndTypeManager, fromKeyType, toKeyType, true);
-        MethodHandle valueProcessor = buildProcessor(functionAndTypeManager, fromValueType, toValueType, false);
+        MethodHandle keyProcessor = (fromKeyType == toKeyType) ? null : buildProcessor(functionAndTypeManager, fromKeyType, toKeyType, true);
```

Comment:
Does it require .equals method or they are the same instance and we can safely use equals method ?

Reply:
yeah thought about that - for primitive types we do have singletons so equals works. I say it's fine for other non-primitive types if the equality is not true.